### PR TITLE
o/snapstate: install snap after prereq snaps are done installing

### PR DESCRIPTION
**Pull request by MiguelPires**
This PR mitigates the issue described in this [forum post](https://forum.snapcraft.io/t/installation-of-snap-with-content-snap-as-dependency/16103). In short, snaps that depend on a content interface and specify a content snap as default-provider are made available to the system before the content snap is installed. Since the content snap might take a bit to install, there's a fair amount of time in which the snap is available but the default-provider it specified isn't installed yet. This is due to the fact that during the installation, the "prerequisite" task schedules tasks to install the content snap but doesn't wait for them to be done. This commit makes the prerequisite's dependants wait for installation of the content snap to be completed. 
In my manual testing, the amount of time the snap is available but failing is reduced from 30 seconds to 1 or 2 seconds. The issue isn't fully resolved because the content snap's interface slot still needs to be auto-connected to the snap's plug and this only happens after the "link-snap" task has made the snap available.

### CODE REVIEW

File: overlord/snapstate/handlers.go [hunk 1]
```diff
@@ -330,6 +330,13 @@ func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string
 		if ts == nil {
 			continue
 		}
+
+		// make the dependencies of the "prerequisite" task wait for the install tasks. Otherwise, the main snap will be
+		// installed in parallel with its prerequisites, resulting in a race condition
+		for _, dep := range t.HaltTasks() {
+			lastPrereqTask := ts.Tasks()[len(ts.Tasks())-1]
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply overlord/snapstate/handlers.go:337]: That's right, the tasks are added to the array in order and chained to wait for the previous task. I'll add a comment then :slightly_smiling_face:

**Comment by MiguelPires** [comment]
Thanks for the review @mardy :slightly_smiling_face:  I improved the comment and name. I also restated the test assertion to be slightly more obvious (imo) and added a check that the task set isn't empty (it shouldn't be but this is more defensive)

File: overlord/snapstate/handlers.go [hunk 2]
```diff
@@ -327,9 +327,17 @@ func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string
...
-		if ts == nil {
+		if ts == nil || len(ts.Tasks()) == 0 {
 			continue
 		}
+
+		// make the dependencies of the "prerequisite" task wait for the install tasks. Otherwise, the main snap will be
+		// installed in parallel with its prerequisites, resulting in a race condition. Waiting for the last task is
+		// enough since tasks wait for the previous one to be done before running
+		for _, dep := range t.HaltTasks() {
+			lastPrereqTask := ts.Tasks()[len(ts.Tasks())-1]
```
Comment (mardy) [inline overlord/snapstate/handlers.go:338]: Sorry for realizing it only now, but this line could actually be moved outside of the loop. This will reduce the calls to `ts.Tasks()`. Then, they could be reduced even more (we would anyway call it three times: two to get the length, and once to extract the last element) by calling it only once and cache its value, but I'm no golang expert so I'm not sure if this is worth it or if the golang compiler is smart enough to optimize away the extra calls.

**Review by mardy — APPROVED** [review]
LGTM, and thanks for adding the comments! :-)

There's one more small possible improvement that I didn't notice before, but it's not critical. :-)

File: overlord/snapstate/handlers.go (hunk 2 above)
Comment (mardy) [reply overlord/snapstate/handlers.go]: Thanks!

File: .gitignore [hunk 3]
```diff
@@ -85,4 +86,4 @@ cmd/test-driver
 *.snap
 
 # Image files
-*.img
\ No newline at end of file
+*.img
```
Comment (stolowski) [inline .gitignore:89]: Unrelated changes should be in a separate PR.

**Review by stolowski — APPROVED** [review]
Thanks, looks good! Even if it doesn't completely fix the problem due to late auto-connect, it is the right thing to do IMO.

One small remark about unrelated .gitignore changes.

File: .gitignore (hunk 3 above)
Comment (MiguelPires) [reply .gitignore:89]: Thanks for the feedback! The commit has already been merged so I'll drop it when I squash the other commits (I included it because it hadn't been merged when I opened the PR)

**Comment by MiguelPires** [comment]
Thanks for the feedback everyone. I spoke to Samuele and it seems like this isn't the right way to go about fixing this issue. A snap's installation will already wait for the default provider's prerequisite task, although this is done later (in the auto-connect task) and in a different way. The auto-connect task checks is there are tasks currently running to install the default provider and, if so, waits a bit before retrying. It also takes into account that there might be circular dependencies and doesn't retry if the prerequisite task is waiting for it complete (which this PR doesn't do as I didn't realise this was possible).
https://github.com/snapcore/snapd/blob/d219fb6b13b88e30df5f0ed077b2c1ade928da27/overlord/ifacestate/handlers.go#L1134

However, since this is done after the snap has been linked, there is a significant period in which snaps can run without their default providers. We will be able to fix this using the inhibit files mechanism to prevent the snap from running before the default provider is ready.
